lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+    end
+    local C = (1 - math.abs(2 * lightness - 1)) * saturation
+    local X = C * (1 - math.abs((hue / 60) % 2 - 1))
+    local m = lightness - (C / 2)
+    local r,g,b = 0,0,0
+    if 0 <= hue and hue <= 60 then
+        r,g,b = C,X,0
+    elseif 60 <= hue and hue <= 120 then
+        r,g,b = X,C,0
+    elseif 120 <= hue and hue <= 180 then
+        r,g,b = 0,C,X
+    elseif 180 <= hue and hue <= 240 then
+        r,g,b = 0,X,C
+    elseif 240 <= hue and hue <= 300 then
+        r,g,b = X,0,C
+    elseif 300 <= hue and hue <= 360 then
+        r,g,b = C,0,X
+    end
+    return unpack({r+m, g+m, b+m})
+end
+
+-- Expects Red, Blue, and Green [0,1]
+-- Does NOT check bounds
+local function rgb_to_hsl(red, green, blue)
+    local h,s,l = 0,0,0 -- luacheck: ignore
+    local M = math.max(red, green, blue)
+    local m = math.min(red, green, blue)
+    local C = M - m
+    -- get Hue
+    if C == 0 then
+        h = 0 -- undefined technically
+    elseif M == red then
+        h = ((green - blue) / C) % 6
+    elseif M == green then
+        h = ((blue - red) / C) + 2
+    elseif M == blue then
+        h = ((red - green) / C) + 4
+    end
+    h = 60 * h
+    -- Lightness/Luminosity
+    l = (M + m) / 2
+    -- Saturation
+    if l == 1 or l == 0 then
+        s = 0
+    else
+        s = C / (1 - math.abs((2 * l) - 1))
+    end
+    return unpack({h, s, l})
+end
+
+-- Create a color object
+local function create_color_object(r,g,b,h,s,l,a)
+    local o = setmetatable({
+        red = r, green = g, blue = b,
+        hue = h, saturation = s, lightness = l,
+        alpha = a
+    }, color_mt)
+    -- allow as functions to use color object variables
+    setmetatable(o.as, {
+        __index = o
+    })
+    return o
+end
+
+-- }}} Helper functions
+
+-- {{{ New Color functions
+
+--- Color object creation functions
+-- @section color_creation
+
+-- {{{ color.new()
+
+--- Create a new color instance
+-- @param val A CSS color compatible string, decimal color value, or a table
+-- containing 3 or 4 values.
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.new({30, 0.8, 0.6})
+-- -- For more examples see the top of this page
+function color.new(val)
+    if type(val) == "string" then return color.from.string(val) end
+    if type(val) == "number" then
+        if val > 0xFFFFFF then return color.from.hex4(val) end
+        return color.from.hex3(val)
+    end
+    if type(val) == "table" then
+        -- do we have red/green/blue/alpha fields?
+        if val.red ~= nil and val.green ~= nil and val.blue ~= nil then
+            local alpha = val.alpha or 255
+            return color.from.rgba(val.red, val.green, val.blue, alpha)
+        -- do we have hue/saturation/lightness/alpha fields?
+        elseif val.hue ~= nil and val.saturation ~= nil and val.lightness ~= nil then
+            local alpha = val.alpha or 1
+            return color.from.hsla(val.hue, val.saturation, val.lightness, alpha)
```

Comment:
So it's a little bit of mix between hexadecimal definition (RGBA) and HSLA is defined, and I wanted to try to keep parity with that as best I could. Internally the numbers will all be `[0,1]` because that's how the RGB to HSL function works (RGB needs to be `[0,1]` and vice versa).